addons/mrp/models/mrp_production.py

```diff
@@ -893,6 +893,9 @@ def write(self, vals):
                 new_date_start = fields.Datetime.to_datetime(vals.get('date_start'))
                 if not production.date_finished or new_date_start >= production.date_finished:
                     production.date_finished = new_date_start + datetime.timedelta(hours=1)
+            if 'picking_type_id' in vals and production.state == 'draft':
+                picking_type = self.env['stock.picking.type'].browse(vals.get('picking_type_id'))
+                production.name = picking_type.sequence_id.next_by_id()
```

Comment:
`production.name = ...` will call again the `write()` method. I would move this chunk above the `super` and update the `vals` dict instead (same for repair order and stock picking) 

Also, could you write a test for the 3 uses cases. 
Thank you